cmd/uhc/cluster/list/list.go

```diff
@@ -0,0 +1,170 @@
+			"parameters or multiple values for the same parameter.",
+	)
+	flags.StringArrayVar(
+		&args.header,
+		"header",
+		nil,
+		"Headers to add to the request. The value must be the name of the header "+
+			"followed by an optional equals sign and then the value of the "+
+			"header. Can be used multiple times to specify multiple headers "+
+			"or multiple values for the same header.",
+	)
+}
+func run(cmd *cobra.Command, argv []string) {
+
+	if len(argv) > 0 {
+		switch argv[0] {
+		case "managed":
+			managed = "managed='true'"
+		case "unmanaged":
+			managed = "managed='false'"
+		default:
+			fmt.Fprintf(os.Stderr, "Unrecognized filter\n")
+			os.Exit(1)
+		}
+	} else {
+		managed = "managed='true'"
+	}
+
+	// Load the configuration file:
+	cfg, err := config.Load()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Can't load config file: %v\n", err)
+		os.Exit(1)
+	}
+	if cfg == nil {
+		fmt.Fprintf(os.Stderr, "Not logged in, run the 'login' command\n")
+		os.Exit(1)
+	}
+
+	// Check that the configuration has credentials or tokens that haven't have expired:
+	armed, err := config.Armed(cfg)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Can't check if tokens have expired: %v\n", err)
+		os.Exit(1)
+	}
+	if !armed {
+		fmt.Fprintf(os.Stderr, "Tokens have expired, run the 'login' command\n")
+		os.Exit(1)
+	}
+
+	// Create the connection:
+	logger, err := util.NewLogger(args.debug)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Can't create logger: %v\n", err)
+		os.Exit(1)
+	}
+
+	// Create the connection, and remember to close it:
+	connection, err := client.NewConnectionBuilder().
+		Logger(logger).
+		TokenURL(cfg.TokenURL).
+		Client(cfg.ClientID, cfg.ClientSecret).
+		Scopes(cfg.Scopes...).
+		URL(cfg.URL).
+		User(cfg.User, cfg.Password).
+		Tokens(cfg.AccessToken, cfg.RefreshToken).
+		Insecure(cfg.Insecure).
+		Build()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Can't create connection: %v\n", err)
+		os.Exit(1)
+	}
+	defer connection.Close()
+
+	// Get the client for the resource that manages the collection of clusters:
+	collection := connection.ClustersMgmt().V1().Clusters()
+
+	// Retrieve the collection of clusters:
+	response, err := collection.List().
+		Search(managed).
+		Send()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Can't retrieve clusters: %s", err)
+		os.Exit(1)
+	}
+
+	// Print the result:
+	response.Items().Each(func(cluster *v1.Cluster) bool {
+		fmt.Printf("%s\n", cluster.Name())
+		return true
+	})
+
+	totalItems := response.Total()
+	sizeOfPage := response.Size()
+	totalPages := totalItems / sizeOfPage
+	for i := 1; i < totalPages; i++ {
+		response, err := collection.List().
+			Search(managed).
+			Send()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Can't retrieve clusters: %s", err)
+			os.Exit(1)
+		}
+		// Print the result:
+		response.Items().Each(func(cluster *v1.Cluster) bool {
+			fmt.Printf("%s\n", cluster.Name())
+			return true
+		})
+	}
```

Comment:
Instead of repeating the code to do the list twice I'd suggest to select an specific page size in advance, and then use a loop to get pages till you get one page that has less results than you requested, as that indicates the last page. For examaple:

```go
pageSize := 10
pageIndex := 1

for {
        // Fetch the next page:
        response, err := collection.List().
                Search(managed).
                Size(pageSize).
                Page(pageINdex).
                Send()
        if err != nil {
                ...
        }

        // Display the fetched page:
        response.Items().Each(func(cluster *v1.Cluster) bool {
                ....
        })

        // If the number of fetched results is less than requested, then
        // this was the last page, otherwise process the next one:
        if response.Size() < pageSize {
                break
        }
        pageIndex++
}
```

Reply:
Thanks, I was trying to figure out a better way to do this